File: #0446.arithmetic-slices-ii-subsequence.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& nums) {

        int size = nums.size();
        vector<unordered_map<long, long>> dp(size);
        int result = 0;
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < i; j++) {
                long diff = (long)nums[i] - (long)nums[j];
                dp[i][diff]++;
                if (dp[j].find(diff) != dp[j].end()) {
                    dp[i][diff] += dp[j][diff];
                    result += dp[j][diff];
                }
            }
        }
        return result;
    }
};

File: #0446.arithmetic-slices-ii-subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var numberOfArithmeticSlices = function(nums) {
    const length = nums.length;
    let dp = new Array(length); 
    let result = 0;
    for (let i = 0; i < length; i++) {
        dp[i] = new Map();
        for (let j = 0; j < i; j++) {
            const diff = nums[i] - nums[j];
            const count = dp[j].get(diff) || 0;
            dp[i].set(diff, (dp[i].get(diff) || 0) + count + 1);
            result += count;
        }
    }
    return result;
};
